### ISSUE TITLE
How to correctly pass a post logout URI from an OIDC client with Code Flow

### ISSUE THREAD

**Issue by andrewbuckingham-waterstons** [DuendeSoftware/Support#107]
**Which version of Duende IdentityServer are you using?**

6.1.1

**Which version of .NET are you using?**

.NET 6

**Describe the bug**

The OIDC client application is using code flow, and passing a `post_logout_redirect_uri` query parameter to IdentityServer in the `/connect/endsession` call, but the resulting `LogoutRequest` from `DefaultIdentityServerInteractionService.GetLogoutContextAsync` returns a null `PostLogoutRedirectUri`

**To Reproduce**

* Set up IdentityServer with ASP.NET Core Identity scaffolding
* Add an in-memory client config with configuration for permitted post logout redirect URIs, and **`authorization_code`** flow
* Set the ASP.NET Core Identity option `LogoutOptions.AutomaticRedirectAfterSignOut = true`
* Configure an OIDC client application (e.g. SPA) and pass a `post_logout_redirect_uri` in the signout redirect call
* Observe that the `LogoutRequest` contains valid data for ClientId, SessionId etc; but the `PostLogoutRedirectUri` is null.

**Expected behavior**

The `LogoutRequest` expected to have the `PostLogoutRedirectUri` as supplied by the OIDC client.

**Log output/exception with stacktrace**

1. OIDC client SPA is calling `endsession` with a `post_logout_redirect_uri` and then the Logout action is triggered:

![image](https://user-images.githubusercontent.com/106542266/176490978-d369de78-2b6f-4e15-8448-370b4bdd0f44.png)

2. IdentityServer logs show that the end session parameters are read, and that a `post_logout_redirect_uri` exists as a parameter, but it parses with a null `PostLogOurUri`:

```
[17:25:45 Debug] Duende.IdentityServer.Validation.EndSessionRequestValidator
Start end session request validation

[17:25:45 Information] Duende.IdentityServer.Validation.EndSessionRequestValidator
End session request validation success
{"ClientId": null, "ClientName": null, "SubjectId": "5918-[redacted]", "PostLogOutUri": null, "State": null, "Raw": {"post_logout_redirect_uri": "http://localhost:3000/"}, "$type": "EndSessionRequestValidationLog"}
```

3. Accordingly, the `LogoutRequest` context shows that all the properties are correctly populated, except for the `PostLogoutRedirectUri` being null:

![image](https://user-images.githubusercontent.com/106542266/176491608-dc2581f1-5f36-4c91-b7c4-9420b879ec6c.png)

**Additional context**

I've noted that the standards recommend the `/connect/endsession` is called with `id_token_hint` when passing a `post_logout_redirect_uri`, but I can't see how this would be necessary with Code Flow (i.e. the client doesn't have a `token_id` so presumably it wouldn't be possible to send one.)

I've also had a look at the `EndSessionRequestValidator` source code in IdentityServer, and can see that the `post_logout_redirect_uri` is not propagated if the `id_token_hint` is null. The code does a check first:

```
if (idTokenHint.IsPresent())
{
   [... assign the post logout redirect]
}
else
{
   [... only assign the Subject, SessionId, ClientIds]
}
```

So, I can't see how the Code Flow OIDC client can request a redirect to a post logout URI. Apologies if this is my lack of understanding. Please could you help?

Many thanks.

**Comment by brockallen** [comment]
> I've noted that the standards recommend the /connect/endsession is called with id_token_hint when passing a post_logout_redirect_uri, but I can't see how this would be necessary with Code Flow (i.e. the client doesn't have a token_id so presumably it wouldn't be possible to send one.)

Yes, to authenticate the call, and thus to know which client is really making the request, the id_token_hint is required. And code flow clients do get an id_token on the response from the token endpoint when they exchange the code. HTH

**Comment by brockallen** [comment]
> I believe id_token isn't provided for Authorization Code flow

Does your client request the "openid" scope? If so, then you should see an id_token on the backchannel response.

**Comment by andrewbuckingham-waterstons** [comment]
That's working now, thank you so much.

For the record, the solution was indeed to retain Authorization Code flow in the client but requesting `openid` as an additional scope in the SPA client config, e.g.:

```
  const oidcConfig: AuthProviderProps = React.useMemo(
    (): AuthProviderProps => ({
      response_type: 'code',
      scope: 'somecustomscope openid',
      post_logout_redirect_uri: 'https://some url in the spa',
```

I then configured IdentityServer to support openid by adding to the builder:
```
.AddInMemoryIdentityResources(new IdentityResource[] { new IdentityResources.OpenId() })
```

After that, `oidc-client-ts` with the `react-oidc-context` React wrappers are correctly passing the token hint and logout URI during `signoutRedirect`.

All good. Thanks again!
